velox/exec/tests/utils/TpchQueryBuilder.cpp

```diff
@@ -751,6 +753,88 @@ TpchPlan TpchQueryBuilder::getQ14Plan() const {
   return context;
 }
 
+TpchPlan TpchQueryBuilder::getQ16Plan() const {
+  std::vector<std::string> partColumns = {
+      "p_brand", "p_type", "p_size", "p_partkey"};
+  std::vector<std::string> supplierColumns = {"s_suppkey", "s_comment"};
+  std::vector<std::string> partsuppColumns = {"ps_partkey", "ps_suppkey"};
+
+  const auto partSelectedRowType = getRowType(kPart, partColumns);
+  const auto& partFileColumns = getFileColumnNames(kPart);
+  const auto supplierSelectedRowType = getRowType(kSupplier, supplierColumns);
+  const auto& supplierFileColumns = getFileColumnNames(kSupplier);
+  const auto partsuppSelectedRowType = getRowType(kPartsupp, partsuppColumns);
+  const auto& partsuppFileColumns = getFileColumnNames(kPartsupp);
+
+  auto planNodeIdGenerator = std::make_shared<PlanNodeIdGenerator>();
+  core::PlanNodeId partScanNodeId;
+  core::PlanNodeId supplierScanNodeId;
+  core::PlanNodeId partsuppScanNodeId;
+
+  auto part = PlanBuilder(planNodeIdGenerator, pool_.get())
+                  .tableScan(
+                      kPart,
+                      partSelectedRowType,
+                      partFileColumns,
+                      {"p_size in (49, 14, 23, 45, 19, 3, 36, 9)"},
+                      "p_type NOT LIKE 'MEDIUM POLISHED%'")
+                  .capturePlanNodeId(partScanNodeId)
+                  // Neq is unsupported as a tableScan subfield filter for
+                  // Parquet source.
+                  .filter("p_brand <> 'Brand#45'")
+                  .planNode();
+
+  auto supplier = PlanBuilder(planNodeIdGenerator)
+                      .tableScan(
+                          kSupplier,
+                          supplierSelectedRowType,
+                          supplierFileColumns,
+                          {},
+                          "s_comment LIKE '%Customer%Complaints%'")
+                      .capturePlanNodeId(supplierScanNodeId)
+                      .planNode();
+
+  auto plan =
+      PlanBuilder(planNodeIdGenerator)
+          .tableScan(kPartsupp, partsuppSelectedRowType, partsuppFileColumns)
+          .capturePlanNodeId(partsuppScanNodeId)
+          .hashJoin(
+              {"ps_partkey"},
+              {"p_partkey"},
+              part,
+              "",
+              {"ps_suppkey", "p_brand", "p_type", "p_size"})
+          .hashJoin(
+              {"ps_suppkey"},
+              {"s_suppkey"},
+              supplier,
+              "",
+              {"ps_suppkey", "p_brand", "p_type", "p_size"},
+              core::JoinType::kAnti)
+          // Empty aggregate is used here to get the distinct count of
+          // ps_suppkey.
+          // approx_distinct could be used instead for getting the count of
+          // distinct ps_suppkey but since approx_distinct is non deterministic
+          // and the standard error can not be set to 0, it is not used here.
+          .localPartition({"ps_suppkey"})
```

Comment:
A typical pattern is to use partial aggregation followed by repartitioning on a subset of group-by keys followed by final aggregation. Is there any particular reason not to use that pattern here?

Also, I'm curious why did we choose {subkey} as a set of columns to partition by. This works as it is a subset of the grouping keys, but I wonder why did we choose this particular subset and not some other one or the full set.

Reply:
`localPartition({"ps_suppkey"}).singleAggregation({"p_brand", "p_type", "p_size", "ps_suppkey"}, {})` is being used here to achieve `distinct ps_suppkey`.
We felt only the `ps_suppkey` as the partition subkey is sufficient to achieve  `distinct ps_suppkey`.
But reasoning this further, I now doubt if the `singleAggregation` will result in `distinct ps_suppkey`, since the grouping keys also contain other columns.
@pramodsatya How is `distinct` mapped from Presto?

Reply:
@majetideepak count(distinct ps.suppkey) can be implemented as a group-by on "p_brand", "p_type", "p_size", "ps_suppkey" with no aggregates followed by group-by on "p_brand", "p_type", "p_size" with count(1). Group by  requires data to be partitioned on a subset of grouping keys, hence, partitioning on just ps_suppkey works. However, by default, one would partition on all grouping keys. Partitioning on a subset keys is useful if data is already pre-partitioned. In this case, it is not necessary to partition again.

```
select
	p.brand,
	p.type,
	p.size,
	count(distinct ps.suppkey) as supplier_cnt
```

Reply:
Presto uses a partial agg followed by final agg for count(distinct). This pattern was not followed here because singleAgg was found to take lesser time than partialAgg + finalAgg.

Reply:
@mbasmanova thank you for clarifying! The `singleAggregation` choice was made based on the performance results.
The % of duplicates < `0.01%`, so singleAggregation was better. But in hind sight, we should stick with a plan that works well in all the cases, so we will follow your suggestion of partial aggregation followed by repartitioning on a subset of group-by keys followed by final aggregation.

Reply:
Thanks, updated the plan accordingly